### PR TITLE
Let `Data` and `HeteroData` implement `FeatureStore`

### PR DESCRIPTION
**Pull request by mananshah99**
This PR lets `Data` and `HeteroData` implement the feature store abstraction. In particular, it defines `put_tensor`, `get_tensor`, and `remove_tensor` methods on both classes, and adds basic tests for these functionalities.

### CODE REVIEW

File: torch_geometric/data/feature_store.py [hunk 1]
```diff
@@ -245,7 +245,7 @@ def __init__(self, attr_cls: Any = TensorAttr):
         attributes by subclassing :class:`TensorAttr` and passing the subclass
         as :obj:`attr_cls`."""
         super().__init__()
-        self._attr_cls = attr_cls
+        self.__dict__['_attr_cls'] = attr_cls
```
Comment (yaoyaowd) [inline torch_geometric/data/feature_store.py:248]: why ?

File: torch_geometric/data/hetero_data.py [hunk 2]
```diff
@@ -616,6 +625,48 @@ def _consistent_size(stores: List[BaseStorage]) -> List[str]:
...
+            if val is not None:
+                val[attr.index] = tensor
+        else:
+            # No group name, just store tensor in node storage:
+            setattr(self[attr.group_name], attr.attr_name, tensor)
+        return True
+
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Obtains a feature tensor from node storage."""
+        # Retrieve tensor and index accordingly:
```
Comment (yaoyaowd) [inline torch_geometric/data/hetero_data.py:648]: In general, should we get `attr.index` first and then `getattr`. The reason I'm asking because it is usually preferred to filter as much data as possible and I think select by index filters more data.

File: torch_geometric/data/data.py [hunk 3]
```diff
@@ -692,6 +712,44 @@ def num_faces(self) -> Optional[int]:
...
+            out[attr.index] = tensor
+        else:
+            # No attr nane, just store tensor:
+            setattr(self, attr.attr_name, tensor)
+        return True
+
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Obtains a feature tensor from node storage."""
+        # Retrieve tensor and index accordingly:
+        tensor = getattr(self, attr.attr_name)
```
Comment (yaoyaowd) [inline torch_geometric/data/data.py:735]: ditto.

File: torch_geometric/data/data.py [hunk 4]
```diff
@@ -365,6 +381,10 @@ def __init__(self, x: OptTensor = None, edge_index: OptTensor = None,
         for key, value in kwargs.items():
             setattr(self, key, value)
 
+        # `Data` does not support group_nae, so we need to adjust `TensorAttr`
```
Comment (rusty1s) [inline torch_geometric/data/data.py:384]:
```suggestion
        # `Data` does not support group_name, so we need to adjust `TensorAttr`
```

File: torch_geometric/data/hetero_data.py [hunk 5]
```diff
@@ -616,6 +625,48 @@ def _consistent_size(stores: List[BaseStorage]) -> List[str]:
 
         return data
 
+    # :obj:`FeatureStore` interface ###########################################
+
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Stores a feature tensor in node storage."""
+        if not attr.is_set('index'):
+            attr.index = None
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:633]: Why do we do need to set `attr.index = None` here?

**Review by rusty1s — COMMENTED** [review]
I think this looks pretty great. Thank you! Let me know if you need further help for the `DynamicInheritance` issue in `Batch`.

On a different note: It looks like `Data` and `FeatureStore` implement, e.g., `__getitem__`. Currently, it looks like we do not really take advantage of the `View` stuff in `FeatureStore` as a result. Might be good to add this as a follow up TODO. WDYT?

File: torch_geometric/data/data.py [hunk 6]
```diff
@@ -692,6 +712,44 @@ def num_faces(self) -> Optional[int]:
...
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Obtains a feature tensor from node storage."""
+        # Retrieve tensor and index accordingly:
+        tensor = getattr(self, attr.attr_name)
+        if tensor is not None:
+            # TODO this behavior is a bit odd, since TensorAttr requires that
+            # we set `index`. So, we assume here that indexing by `None` is
+            # equivalent to not indexing at all, which is not in line with
+            # Python semantics.
+            return tensor[attr.index] if attr.index is not None else tensor
```
Comment (rusty1s) [inline torch_geometric/data/data.py:741]: One important thing: I think we need to make use of `cat_dim` here to know which in which index to select. See [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/loader/utils.py#L103-L104) for an example:
```python
cat_dim = self.__cat_dim__(attr.attr_name, tensor)
```

File: torch_geometric/data/data.py [hunk 7]
```diff
@@ -365,6 +381,10 @@ def __init__(self, x: OptTensor = None, edge_index: OptTensor = None,
         for key, value in kwargs.items():
             setattr(self, key, value)
 
+        # `Data` does not support group_nae, so we need to adjust `TensorAttr`
+        # accordingly here to avoid requiring `group_name` to be set:
+        FeatureStore.__init__(self, attr_cls=DataTensorAttr)
```
[the remark this answers is not in the document]
Comment (mananshah99) [reply torch_geometric/data/data.py:386]: Or we could just take advantage of MRO and call `super(Data, self).__init__(attr_cls)`; ended up implementing that way as I think it's cleaner. lmkwyt :)

File: torch_geometric/data/data.py [hunk 8]
```diff
@@ -692,6 +712,44 @@ def num_faces(self) -> Optional[int]:
...
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
+        r"""Stores a feature tensor in node storage."""
+
+        if not attr.is_set('index'):
+            attr.index = None
+
+        out = getattr(self, attr.attr_name, None)
+        if out is not None:
+            # Attr name exists, handle index:
+            out[attr.index] = tensor
```
[the remark this answers is not in the document]
Comment (mananshah99) [reply torch_geometric/data/data.py:726]: Good catch :)

File: torch_geometric/data/data.py (hunk 6 above)
Comment (mananshah99) [reply torch_geometric/data/data.py:741]: I imagined that it would be on the user to provide the correct slice for this; for example,
```
(*([slice(None)] * (cat_dim - 1 if cat_dim > 0 else cat_dim)), attr.index)
```
but I'm happy to implement it this way as well.

File: torch_geometric/data/hetero_data.py (hunk 2 above)
Comment (mananshah99) [reply torch_geometric/data/hetero_data.py:646]: Not sure I understand. In this implementation, it doesn't matter since all data is in memory, so we first need to get the tensor and then index that tensor. In an out-of-core implementation, all of `TensorAttr` is passed to the implementor, and they can filter however they like.

File: torch_geometric/data/data.py (hunk 6 above)
Comment (rusty1s) [reply torch_geometric/data/data.py]: I see. That makes sense.

File: torch_geometric/data/hetero_data.py [hunk 9]
```diff
@@ -616,6 +623,50 @@ def _consistent_size(stores: List[BaseStorage]) -> List[str]:
 
         return data
 
+    # :obj:`FeatureStore` interface ###########################################
+
+    def _put_tensor(self, tensor: FeatureTensorType, attr: TensorAttr) -> bool:
```
Comment (yaoyaowd) [inline torch_geometric/data/hetero_data.py:628]: Do we really want to support this `put_tensor` operation now? It looks like there are a lot of places could go wrong. What if index is out of bound (put new tensors)? What if `out` does not exist and `attr.index` is not continuous? What if `out` exists but `val` does not? I'd rather throw a not implemented error here.

**Review by yaoyaowd — APPROVED** [review]
Minor comment, LGTM.

File: torch_geometric/data/hetero_data.py [hunk 10]
```diff
@@ -616,6 +623,50 @@ def _consistent_size(stores: List[BaseStorage]) -> List[str]:
...
+        else:
+            # No node storage found, just store tensor in new one:
+            setattr(self[attr.group_name], attr.attr_name, tensor)
+        return True
+
+    def _get_tensor(self, attr: TensorAttr) -> Optional[FeatureTensorType]:
+        r"""Obtains a feature tensor from node storage."""
+        # Retrieve tensor and index accordingly:
+        tensor = getattr(self[attr.group_name], attr.attr_name)
+        if tensor is not None:
```
Comment (Padarn) [inline torch_geometric/data/hetero_data.py:648]: If it requires we set `index` why do we even need to check?

Comment (Padarn) [reply torch_geometric/data/hetero_data.py:648]: Maybe another way to solve this if its confusing is to have a different value for `UNSET` which indicates its going to index 'all'

File: torch_geometric/data/batch.py [hunk 11]
```diff
@@ -23,8 +23,12 @@ def __call__(cls, *args, **kwargs):
             new_cls = base_cls
         else:
             name = f'{base_cls.__name__}{cls.__name__}'
+
+            class MetaResolver(type(cls), type(base_cls)):
```
Comment (rusty1s) [inline torch_geometric/data/batch.py:27]: Can we add a short description to why this is necessary?

File: torch_geometric/data/data.py [hunk 12]
```diff
@@ -348,7 +364,10 @@ class Data(BaseData):
     def __init__(self, x: OptTensor = None, edge_index: OptTensor = None,
                  edge_attr: OptTensor = None, y: OptTensor = None,
                  pos: OptTensor = None, **kwargs):
-        super().__init__()
+        # `Data` doesn't support group_name, so we need to adjust `TensorAttr`
+        # accordingly here to avoid requiring `group_name` to be set:
+        super(Data, self).__init__(attr_cls=DataTensorAttr)
```
Comment (rusty1s) [inline torch_geometric/data/data.py:369]:
```suggestion
        super().__init__(attr_cls=DataTensorAttr)
```
should work as well?

File: torch_geometric/data/data.py [hunk 13]
```diff
@@ -692,6 +714,45 @@ def num_faces(self) -> Optional[int]:
             return self.face.size(self.__cat_dim__('face', self.face))
         return None
 
+    # FeatureStore interface ###########################################
+
+    def items(self):
+        return self._store.items()
```
Comment (rusty1s) [inline torch_geometric/data/data.py:720]: Add type hint + doc-string. Why is this necessary for implementing the `FeatureStore` interface?

File: torch_geometric/data/hetero_data.py [hunk 14]
```diff
@@ -92,6 +97,8 @@ class HeteroData(BaseData):
     DEFAULT_REL = 'to'
 
     def __init__(self, _mapping: Optional[Dict[str, Any]] = None, **kwargs):
+        super(HeteroData, self).__init__()
```
Comment (rusty1s) [inline torch_geometric/data/hetero_data.py:100]:
```suggestion
        super().__init__()
```

**Review by rusty1s — APPROVED** [review]
I think this looks great. Thank you! One thing that would be good to confirm is whether pickling of `data` objects between PyG 2.0.4 and PyG master still works, e.g., running an example in PyG 2.0.4 and re-run with pre-processed dataset in master. Otherwise, we have to list this as a breaking change.

File: torch_geometric/data/hetero_data.py (hunk 9 above)
Comment (mananshah99) [reply torch_geometric/data/hetero_data.py:628]: I think we should support `put_tensor` since all of the cases you mentioned (and others) are either covered by the function or error out in the same way that PyG would normally error out. In particular,

> What if index is out of bound (put new tensors)?

Not sure what you mean here. If we are putting a new tensor, then no node storage would be found and we would just store the tensor as intended. If you mean an index out of bounds exception for replacing a tensor, PyG would fail here as well (as expected)

> What if out does not exist and attr.index is not continuous? 

This is fine, we support any index that torch supports

> What if out exists but val does not?

Good point! Fixed.

File: torch_geometric/data/data.py (hunk 13 above)
Comment (mananshah99) [reply torch_geometric/data/data.py:720]: I spent far too long on this; it's because `Data` is now a `MutableMapping` (inherited from `FeatureStore`), which defines an `items` method, so we now have to define `items` explicitly. I'll add a docstring to this effect; the end result is that the behavior remains the same.

File: torch_geometric/data/hetero_data.py (hunk 10 above)
Comment (mananshah99) [reply torch_geometric/data/hetero_data.py:648]: Thanks for the comment! Not sure what you mean by needing to check `index`; this is because `TensorAttr` just requires that its attributes are set (they can be set to None). The current way to index all would be `None`, which I think is acceptable; although I'm happy to define a custom value for the `UNSET` enum to indicate `all` indexing in a follow-up PR.
